zio-http/src/main/scala/zio/http/DnsResolver.scala

```diff
@@ -0,0 +1,285 @@
+                        val newEntry = CacheEntry(
+                          Promise.unsafe.make(fiberId)(Unsafe.unsafe),
+                          Some(addresses),
+                          now,
+                        )
+                        startResolvingHost(host, newEntry.resolvedAddresses)
+                          .as(Some(CachePatch.Update(host, newEntry)))
+                      } else {
+                        ZIO.some(CachePatch.Remove(host))
+                      }
+                    } else {
+                      // successful entry not expired yet
+                      ZIO.none
+                    }
+                  },
+                )
+              case None                       =>
+                ZIO.none
+            }
+          }
+          .map(_.flatten)
+      }
+
+    private def ensureMaxSize(entries: Map[String, CacheEntry]): ZIO[Any, Nothing, Chunk[CachePatch]] =
+      if (entries.size > maxCount) {
+        val toDrop = Chunk.fromIterable(entries).sortBy(_._2.lastUpdatedAt).take(entries.size - maxCount)
+        ZIO
+          .foreach(toDrop) { case (host, entry) =>
+            entry.resolvedAddresses.interrupt.as(CachePatch.Remove(host))
+          }
+      } else {
+        ZIO.succeed(Chunk.empty)
+      }
+
+    private def applyPatches(entries: Map[String, CacheEntry], patches: Chunk[CachePatch]): Map[String, CacheEntry] =
+      patches.foldLeft(entries) { case (entries, patch) =>
+        patch match {
+          case CachePatch.Remove(host)               =>
+            entries - host
+          case CachePatch.Update(host, updatedEntry) =>
+            entries.updated(host, updatedEntry)
+        }
+      }
+  }
+
+  private object CachingResolver {
+    def make(
+      resolver: DnsResolver,
+      ttl: Duration,
+      unknownHostTtl: Duration,
+      maxCount: Int,
+      maxConcurrentResolutions: Int,
+      expireAction: ExpireAction,
+      refreshRate: Duration,
+    ): ZIO[Scope, Nothing, DnsResolver] =
+      for {
+        semaphore <- Semaphore.make(maxConcurrentResolutions)
+        entries   <- Ref.make(Map.empty[String, CacheEntry])
+        cachingResolver = new CachingResolver(resolver, ttl, unknownHostTtl, maxCount, expireAction, semaphore, entries)
+        _ <- cachingResolver.refreshAndCleanup().scheduleFork(Schedule.fixed(refreshRate))
+      } yield cachingResolver
+  }
+
+  private[http] def snapshot(): ZIO[DnsResolver, Nothing, Map[String, CacheEntry]] =
+    ZIO.service[DnsResolver].flatMap {
+      case cachingResolver: CachingResolver => cachingResolver.snapshot()
+      case _ => ZIO.dieMessage(s"Unexpected DnsResolver implementation: ${getClass.getName}")
+    }
+
+  final case class DnsResolverConfig(
+    ttl: Duration,
+    unknownHostTtl: Duration,
+    maxCount: Int,
+    maxConcurrentResolutions: Int,
+    expireAction: ExpireAction,
+    refreshRate: Duration,
+  )
+
+  object DnsResolverConfig {
```

Comment:
I would probably just call this `Config` since it's already located inside `DnsResolver`, although I realize that makes it necessary to qualify `zio.Config`.

Compare:

```scala
DnsResolver.DnsResolverConfig
```

vs

```scala
DnsResolver.Config
```

Reply:
Yes, but we have https://github.com/zio/zio-http/blob/main/zio-http/src/main/scala/zio/http/ClientConfig.scala etc (although not in the companion object of `Client` - one more thing to unify)